File: android/app/src/main/kotlin/com/example/perfect_corp_homework/api/MethodChannelResponse.kt
package com.example.perfect_corp_homework.api

import com.example.perfect_corp_homework.exception.BadRequestError
import com.example.perfect_corp_homework.exception.JsonSerializationError
import com.example.perfect_corp_homework.exception.NoInternetError
import com.example.perfect_corp_homework.exception.UnSupportImageTypeError
import com.example.perfect_corp_homework.exception.UnknownError

class MethodChannelResponse<T>(serviceResult: ServiceResult<T>) {

    val statusCode: Int = when(serviceResult.error) {
        is UnSupportImageTypeError -> 10000
        is BadRequestError -> 10001
        is NoInternetError -> 10002
        is UnknownError -> 10003
        is JsonSerializationError -> 10004
        is Exception -> 10003
        else -> 0
    }
    val errorMessage: String? = serviceResult.error?.message
    val data: T? = serviceResult.data
}

File: plugin/download/ios/Classes/api/MethodChannelResponse.swift
//
//  MethodChannelResponse.swift
//  Runner
//
//  Created by SamLS Chen on 2024/8/20.
//

import Foundation


class MethodChannelResponse<T: Encodable>{
    
    let statusCode: Int
    let errorMessage: String?
    let data: T?
    
    init (serviceResult: ServiceResult<T>) {
        guard serviceResult.error == nil else {
            switch(serviceResult.error!) {
            case .UnsupportedMediaTypeError:
                statusCode = 10000
            case .BadRequestError:
                statusCode = 10001
            case .UnknownError(_):
                statusCode = 10003
            }
            data = nil
            errorMessage = serviceResult.error!.message
            return
        }
        
        statusCode = 0
        data = serviceResult.data
        errorMessage = ""
    }
}


extension MethodChannelResponse: Encodable {
    enum CodingKeys: String, CodingKey {
        case statusCode
        case errorMessage
        case data
    }
    
    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try? container.encode(statusCode, forKey: .statusCode)
        try? container.encode(errorMessage, forKey: .errorMessage)
        try? container.encode(data, forKey: .data)
    }
}
